### ISSUE TITLE
[no-misused-promises] Promise<void> isn't accepted where definition expects void

### ISSUE THREAD

**Issue by andreineculau** [typescript-eslint/typescript-eslint#1637]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-misused-promises": "error"
  }
}
```

```TS
export type CallbackFn = (p: number) => void;
let withcb = function(cb: CallbackFn): void { };

withcb(async function(): Promise<void> {
  await Promise.resolve();
});
```

**Expected Result**

no complaints

**Actual Result**

complains with "Promise returned in function argument where a void return was expected."

**Additional Info**

Typescript compiler is fine with it.

Online discussion at https://twitter.com/TitianCernicova/status/1232236426193928192

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.20.0` |
| `@typescript-eslint/parser`        | `2.20.0` |
| `TypeScript`                       | `3.8.2` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `13.8.0` |
| `npm`                              | `6.13.7` |

**Comment by bradzacher** [comment]
> Typescript compiler is fine with it.

The typescript compiler is also fine with `if (Promise.resolve(true)) {}`. That doesn't make it correct code.

---

If you check out the rule docs, there's an option that will do what you want.
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-misused-promises.md#options

**Comment by andreineculau** [comment]
I hope that you didn't close the issue just because of a fault in logic in a side-argument. Put differently, even if Typescript compiler is fine with X, doesn't mean that the compiler isn't broken to begin with, thus my argument was crap to begin with. Anyways, forget for a second about the typescript compiler.

Given a callback pattern, the callback caller should not stop working if a callback crashes. Not saying that all implementations in the wild are doing that, but I'm hoping that we agree up to that point.

Given a queue of callbacks for an event for example,

a sound implementation would look more like
```js
callbacks.forEach(function(cb) {
  setImmediate(async function() {
    try {
      await cb(...args);
    } catch (err) {
      console.error(err);
    }
  });
});
```

while this is a broken implementation

```js
callbacks.forEach(function(cb) {
  cb(...args);
});

// broken because
// if cb is sync
// a. it means the n-th cb function will be called late
// b. it means the n-th cb function might never be called if a previous one throws
// c. it means code after the forEach loop never executes if a cb function throws
// if cb is async, b + c still happen, but not a
```

Do you understand now where I'm coming from?
No matter if you understand/agree with me or not, there's a problem though. Eslint cannot differentiate between a case of a callback situation, and a general situation of "expected void, but got Promise<void>".

---

I don't know what is the proper forum to ask questions, so I throw one here: what is that you do in situations where a callback needs to call async code?

```js
emitter.on('some-event', function() {
  let response = await fetch('/some/thing');
});
```

1. make the callback async, and ignore this rule --- a bad choice unless you know that the emitter has a sound implementation for calling the callbacks
2. implement a helper function - an IIFE wrapper with a try-catch like in the "sound" example above
2a. come to think about it, even when we're not dealing with the need for an async callback, you might still want the helper function, just to be sure that if your callback throws, the event emitter implementation doesn't crash
3. ?

thanks in advance!

**Comment by bradzacher** [comment]
> I hope that you didn't close the issue just because of a fault in logic in a side-argument

Nope, I closed it because you can turn off this check via the rule's options
`checksVoidReturn: boolean`
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-misused-promises.md#options

> what is that you do in situations where a callback needs to call async code?

The big problem with passing an async function into a non-async receiver is that you're firing off async work and nothing is watching for its completion or failure. Chances are you've already handled the success case as part of your function, so it's really just the error case that you need to be sure you're explicitly handling.

One other way to work around this by using promises to eliminate the async/await:
```
emitter.on('some-event', function() {
  async function asyncWorker() {
    let response = await fetch('/some/thing');
  }

  asyncWorker().then(handleSuccess, handleError);
});
```

Another option is to just keep using this lint rule as is, and then using an eslint-disable comment once you know you've explicitly handled the error case.
That way the lint rule will remind you to handle the error explicitly, and you can provide a comment which assures reviewers that what you've done is properly handled.

**Comment by andreineculau** [comment]
> checksVoidReturn

🤦‍♂ git history shows me that it has always been there. I don't remember seeing it, and I made the mistake of not reading the docs before opening this issue. I admit, I'm guilty.
But beyond this mistake, it is not an option I will easily disable...

I agree with you top to bottom, and I can see myself going with the boilerplate but given how often this happens, maybe I wouldn't do it happily again and again and then it's also not smth that will be easy to swallow by devs that are not taking deep dives into js.
*Is that really how you/your team(s) are regularly doing the right thing when facing a no-misused-promises complaint?* I'm questioning here not whether to do it, but the long-discussions/read-write-convenience side of it, vs a generic helper function (?! handle exceptions via console.error and shut up) vs disabling for void returns vs disabling the entire rule, etc :)

PS: now that I've fully digested the situation, one has to wonder why did the rule even get a `checksVoidReturn` option in the first place. Seems to hide a great deal of ticking bombs.

PPS: Once again, thanks a bunch! If you find yourself in Stockholm, Sweden and in need of a guide, I'll be happy to pay back my intellectual debt to you :) Cheers!

**Comment by bradzacher** [comment]
> one has to wonder why did the rule even get a checksVoidReturn option in the first place

Yeah it does, but some people don't care. There is a large number of people who don't care about 100% correctness, and prefer developer speed and such.

> Is that really how you/your team(s) are regularly doing the right thing when facing a no-misused-promises complaint?

I work at facebook, so I don't use the plugin or even typescript at work (we use flow). I just maintain this project in my spare time 😄 

[In flow this is actually a compiler error](https://flow.org/try/#0PQKgBAAgZgNg9gdzAZwC4CcCWBjVYTABQhAJgKbYwCG6ZYUArgHa6ZxNhXbZkAOqyAMIAjABTZhALjCiAlGAC8APjAA3OJhKzp6zQG5iwYGAAK6OAFtMyMgB5dJFdbBM4eKsmSYA5kyrCYOlQ4NQ0SQi4efiExOUUVM0trMgA6WmQ4GFUyOVkDIzAAORDVKhgGMmROWk5PHz8AoJKwiO4+ARFROOUwAEY81qiO2PkejAqByPaYrtGVAHI0LCZveYGC1zwybI4mBhgYQenO7pU9g4HCAvYYAE8wZnIoTCYyEjBnDy9ff0CwYNCmiO0ROcweTCeLzeAyAA) - `Promise<void>` is not assignable to `void`, so this isn't a situation that can happen in the facebook codebase.

On top of that, pretty much everything I've worked on has been react with relay for the data layer, so you rarely even have to await anything, so this situation is really a non-issue.

I just had a look, and in the rare cases that an await is required, it's actually built into the component APIs.
I.e. we have a modal component which takes `onConfirm: () => (Promise<void> | void)`, and the  code actually has handling for the promise (it keeps the modal open with a spinner until the promise resolves/rejects).

We don't use very many 3rd party libs, so there's not usually an issue with module boundaries, but it looks like a lot of people just go with the promise approach.
